machines: Move from direct machines.json manipulation to D-Bus calls, move to /etc/cockpit/machines/*.json

martinpitt opened the pull request:
The first two commits implement the first item in the [Finalize machines.json format](https://trello.com/c/WSEOANNY/268-finalize-machines-json-format) trello card. I. e. it keeps `/var/lib/machines.json` for now, but moves the parsing/updating of that from the web frontend (cockpit.file) to a D-Bus interface.

The third commit moves the single `/var/lib/machines.json` to `/etc/cockpit/machines/*.json`. I kept that separately as it was historically developed that way, but it also makes it easier to separately review the move to D-Bus and the move to multiple files.

- [x] glib 2.44 is not available on all targets yet, rewrite `g_auto*` stuff to use manual cleanup
- [x] fix integration test regressions
- [x] drop XXX debug messages
- [x] fix memleaks (aside from [leak in json-glib](https://bugzilla.gnome.org/show_bug.cgi?id=778674), added suppression )
- [x] add unit tests for new D-Bus interface
- [x] move to `/etc/cockpit/machines/*.json`
- [x] fix Update() call to only write the delta
- [x] change Update() call to take the destination file name instead of hardcoding it
- [x] tests for multiple machines/*.json files
- [x] update documentation
- [x] try to factorize `merge_config()`, the outer and inner loop are structurally similar
- [x] aggregate bursts of inotify events into one parse/property update callback

dperpeet commented:
I suggest using a different prefix than `machines` - we have a package with that name that works with virtual machines.

martinpitt commented:
@dperpeet: good point; what about `hosts`? By "prefix" I assume you mean the D-Bus object/interfaces and the new source files.

martinpitt commented:
At least Debian 8 has an older glib (2.42), and the centos-7 build is unhappy too (although there's no useful error message). So I suppose I'll go back to manual unreffing for now. Too bad, this is quite nice :-)

martinpitt commented:
Next version: I reverted the glib version bump and did the unreffing manually, and fixed the `JsonNode` reffing/counting to fix the test regressions (test suite now passes for me locally). But I'll only tick the task list box for that once the automatic results confirm.

martinpitt commented:
note keeping: apparently on debian-8 and centos-7 json-glib is rather old, and does not offer some important API yet: I need to see how to get around that. This is hidden from the CI build logs, so putting it here:
```
src/bridge/cockpitdbusmachines.c:62:5: error: implicit declaration of function 'json_node_ref' [-Werror=implicit-function-declaration]
     result = json_node_ref (result);
     ^
src/bridge/cockpitdbusmachines.c:62:12: warning: assignment makes pointer from integer without a cast
     result = json_node_ref (result);
            ^
src/bridge/cockpitdbusmachines.c: In function 'build_machines':
src/bridge/cockpitdbusmachines.c:77:3: error: implicit declaration of function 'json_node_unref' [-Werror=implicit-function-declaration]
   json_node_unref (machines);
   ^
src/bridge/cockpitdbusmachines.c: In function 'update_machine':
src/bridge/cockpitdbusmachines.c:104:7: error: unknown type name 'JsonObjectIter'
       JsonObjectIter iter;
       ^
src/bridge/cockpitdbusmachines.c:110:7: error: implicit declaration of function 'json_object_iter_init' [-Werror=implicit-function-declaration]
       json_object_iter_init (&iter, json_node_get_object (info));
       ^
src/bridge/cockpitdbusmachines.c:111:7: error: implicit declaration of function 'json_object_iter_next' [-Werror=implicit-function-declaration]
       while (json_object_iter_next (&iter, &prop_name, &prop_node))
       ^
cc1: some warnings being treated as errors
Makefile:4107: recipe for target 'src/bridge/libcockpit_bridge_a-cockpitdbusmachines.o' failed
```

martinpitt commented:
The Debianish tests fail because there `/var/lib/cockpit` is root:root 755 and thus the bridge of an admin user can't write to it. This works in RPMish environments as there the directory is root:wheel 775. While that should be fixed, I don't think we can/should rely on that, so we need a similar "run this through a root bridge" facility than the `"superuser": "try"` flag we used for `cockpit.file`.

martinpitt commented:
I now pushed the "move to /etc/cockpit/machines.d/*.json" changes and fixed the worst integration test fallout. Let's see what remains.

martinpitt commented:
I fixed the remaining issues, and I believe this is now ready for reviewing. All integration tests succeed for me locally, but on the infra I keep getting (unrelated) failures from the usual suspects.

petervo commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+  if (G_UNLIKELY (res != 0 && res != GLOB_NOMATCH))
+    g_error ("glob %s failed with return code %i", glob_str, res);
+  /* also read /var/lib/cockpit/machines.json for backwards compat */
+  conf_glob.gl_pathv[0] = g_build_filename (g_getenv ("COCKPIT_DATA_DIR") ?: "/var/lib/cockpit", "machines.json", NULL);
```
Should we migrate this file once on start up and not have to read it every time.

martinpitt replied:
Migrating the file is part of https://github.com/cockpit-project/cockpit/wiki/Config-format-for-known-machines-and-ssh-keys, this PR just does not implement this (yet). I wasn't completely sure if we can just do that, but I suppose we don't need to be concerned about downgrading -bridge after the migration?

However, it might happen that the migration fails because `/etc/cockpit/machines.d/` cannot be created or written to for some reason. So we need to keep the code that reads `/var/` for some time at least.

This seemed hairy enough to be in a separate PR (this doesn't address known_hosts either yet), but if you prefer I'll fold it in here.

petervo replied:
Seems to me that migrating the file can be done separately from anything to do with known hosts.   But yes, handling the no permissions case properly is a concern. I'm ok with leaving this for a follow up or a remotectl command.

petervo replied:
Just mainly want to make sure we've thought about the cli use cases before finalizing the dbus call params.

martinpitt replied:
I created a work item for that in the Trello card.

petervo commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+    }
+  g_free (path);
+
+  machines = build_machines ();
```
It would be nice to not have to rebuild this from scratch everytime.

martinpitt replied:
This is the callback when `/etc/cockpit/machines.d/` changes, so here we have to rebuild it. Parsing a few small JSON files is rather cheap, I don't want to introduce complex heuristics about which files we don't need to re-parse and keep a per-file partial struct. This would be very involved and error prone.

petervo replied:
I see the issue, i'm just a little concerned about performance if it goes beyond just a couple files. WDYT about adding some throttling to minimize the number of times it gets run or at least ensure that we don't have multiple runs going at the same time?

martinpitt replied:
See [measurements below](https://github.com/cockpit-project/cockpit/pull/5880#issuecomment-281194632). I. e. the thing to optimize is the number of D-Bus calls/property updates, not the parsing. But as we want a reactive UI, I don't think we can just ignore updates for some time after one update got sent, as we can't predict whether the files will change again after that dead time.

petervo replied:
It wouldn't be ignoring events, just batching events so that it's not one run per each event when there are multiple events received in quick succession.

petervo commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return build_machines ();
```
I don't think we want to be rebuilding this on every property call. We should probably mantain a cache of what we've loaded. That might also help with the post monitoring actions.

martinpitt replied:
Agreed, `machines_get_property()` can use a cache as we have an inotify watch in place. Will work on this tomorrow.

petervo commented on pkg/lib/machines.js:
```diff
@@ -176,28 +172,23 @@
+
+            // FIXME: investigate re-using the proxy from Loader (runs in different frame/scope)
+            var bridge = cockpit.dbus(null, { bus: "internal", "superuser": "try" });
+            var mod = bridge.call("/machines", "cockpit.Machines", "Update", [ "99-webui.json", host, values_variant ])
```
I'm not 100% convinced by having the dashboard write to only one file. But if everyone else in on board with that I'm ok with it.

martinpitt replied:
I wondered about that as well in [the recent email thread](https://lists.fedorahosted.org/archives/list/cockpit-devel@lists.fedorahosted.org/thread/NEHK6KNRTZTDEDCNHWTDGASHJBUL6D7P/) ([direct link to that message](https://lists.fedorahosted.org/archives/list/cockpit-devel@lists.fedorahosted.org/message/R5C2GFUCCVURI6TXRXCLEMS5WBHKZSOO/)) .

@mvollmer convinced me that it's better to write a dashboard specific file than to modify the original files, as that might conflict with files shipped by packages or managed by puppet & friends.

stefwalter replied:
That's a pretty good reason. And I think one that we can live with. 

A bigger decision is the implementation of multiple machines per file, rather than enforcing one machine per file. However I think you've made the right call here too.

petervo commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
```
Were you thinking the corresponding remotectl command for this autogenerate the file name? Or expect that as user input. Or should we be doing some autogeneration here?

martinpitt replied:
I thought ATM remotectl only managed the certificate, not `machines.json` entries? Adding a CLI is indeed [a work item in the Trello card](https://trello.com/c/WSEOANNY/268-finalize-machines-json-format) but hasn't been discussed/designed yet. My first version had the `99-webui.json` file name hardcoded in the bridge, but I changed it into an explicit D-Bus argument so that it only needs to be hardcoded in the frontend and thus the D-Bus API is flexible to write other files, for a CLI API to supply the "pre-seeded" files.

I think a CLI for adding a machine should encourage to specify a file name as well, maybe defaulting to "10-<hostname>.json"?

petervo replied:
remotectl currently only manages certs but long term we want to be able to add machines with it. Having a 10-{{hostname}}.json default sounds good to me.

petervo replied:
Just mainly want to make sure we've thought about the cli use cases before finalizing the dbus call params.

martinpitt commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+{
+  static gchar *path = NULL;
+  if (path == NULL)
+    path = g_build_filename (g_getenv ("COCKPIT_CONFIG_DIR") ?: "/etc/cockpit", "machines.d", NULL);
```
Thanks, fixed locally in my branch.

martinpitt commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,416 @@
+
+  path = g_file_get_path (file);
+  if (g_str_has_suffix (path, ".json"))
+    g_debug ("on_machines_changed: event type %i on %s", event_type, path);
```
Fixed in my local branch.

martinpitt commented:
Wrt. performance concerns: I did some timing with 20 files (the maximum number of hosts that we allow):

```C
  struct timespec before, after;
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &before);
  for (unsigned i = 0; i < 10000; ++i)
  {
      GVariant *v = build_machines ();
      g_variant_unref (v);
  }
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &after);
  gint64 delta = (after.tv_sec - before.tv_sec) * 1000000000 + (after.tv_nsec - before.tv_nsec);
  g_warning("time: %" G_GINT64_FORMAT " ns\n", delta);
```

```sh
$ for i in `seq 20`; do echo "{\"host$i\": {\"address\": \"1.2.3.4\"}}" | sudo tee /etc/cockpit/machines.d/01-host$i.json; done
```

which takes between 3.42 and 3.46 s on my system (so the clock is fairly reliable, ran it ten times). I. e. one update takes about 0.34 ms. Even on a slow ARM system it shouldn't take more than a few ms. Also, we don't run concurrent threads, so these don't run simultaneously.

So I don't think we need to over-optimize this. I'll still add a cache for get_property(), but the overhead of funneling this through D-Bus and into JS will outweigh the actual parsing by a lot.

In practical terms these files won't change a lot anyway, and the dashboard only ever reads the property once (at program start) and then just listens to inotify changes.

petervo commented:
That limit is not going to hold in the future, I don't think we should be basing decisions on that limit.

It's not simultaneous runs that i'm concerned about as much as the function just being queued to run again and again in quick succession. Think of a cron or CM process that comes in and regularly touches stuff. Or even a simple ```rm *-hostname.json``` and there are 10 files that match. That will result in this function running 10x.

martinpitt commented:
Indeed, I'll aggregate bursts of inotify events into just one parsing/D-Bus property update. Added a TODO item into the description.

martinpitt commented:
I pushed an update which now aggregates bursts of inotify events into just one round of parsing *.json and sending `PropertiesChanged`, and fixes the little things Peter found.

Notably it does not yet cache the machines property if you do an explicit Get() call -- I have a patch for this, but it is broken both in terms of tests (as a Get call now gets an obsolete value, one needs to retry a few times) and in terms of valgrind (as keeping proper refcounts is a bit ugly), and IMHO it also doesn't buy much really: Our UI only calls Get() once ever (when the bridge is started), and I'm not sure whether this is actually the right thing to do for the API in general -- if you write a file and then call Get(), it's not completely unreasonable to expect that you get the updated properties. With caching you would need to wait until the delayed inotify update kicks in. I keep this open as a todo item above for now.

petervo commented:
I think that makes sense for now.

@stefwalter do you have any comments on the API, specifically having all hosts as one property as well as dashboard only writing to one file.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+/* counts the number of file change events that have not yet gotten a PropertiesChanged signal*/
+static guint pending_updates;
+
+static const char*
```
Style: This should have a space before ```*```

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+}
+
+static int
+glob_err_func (const char *epath, int eerrno)
```
Style: Various arguments are on their own lines prefix with spaces as indentation.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+  return n;
+}
+
+static JsonNode*
```
Style: space before asterisk

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+    {
+      result = json_parser_get_root (parser);
+      /* root is NULL if the file is empty */
+      if (result != NULL) {
```
Style: brace goes on a new line.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+      result = json_parser_get_root (parser);
+      /* root is NULL if the file is empty */
+      if (result != NULL) {
+        if (JSON_NODE_HOLDS_OBJECT (result))
```
Style: If else block has brace, so should the if block.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+          result = json_node_copy (result);
+        else
+          {
+            g_warning ("%s does not contain a JSON object, ignoring", path);
```
Suggest putting a colon after the ```%s``` similar to other warnings elsewhere.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+{
+  /* just log the error for debugging */
+  if (eerrno != ENOENT)
+    g_warning ("Cannot read %s: %s", epath, strerror (eerrno));
```
Suggest using similar format to other warnings: ```g_warning ("%s: cannot read: %s", epath, g_strerror (eerno));```

In addition, note use of g_strerror().

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+  else
+    {
+      if (error->code != G_FILE_ERROR_NOENT)
+        g_warning("%s has invalid JSON: %s", path, error->message);
```
Suggest colon after %s

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+}
+
+static void
+merge_config (JsonObject *machines, JsonObject *delta, const char *path)
```
Style: Arguments should be on different lines.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+        json_object_set_member (machines, hostname, new_object_node ());
+      JsonObject *machine_props = json_object_get_object_member (machines, hostname);
+
+      g_debug("%s: merging updates to host name %s", path, hostname);
```
Style: space after g_debug().

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+  glob_str = g_build_filename (machines_json_dir(), "*.json", NULL);
+  res = glob (glob_str, GLOB_DOOFFS, glob_err_func, &conf_glob);
+  if (G_UNLIKELY (res != 0 && res != GLOB_NOMATCH))
+    g_error ("glob %s failed with return code %i", glob_str, res);
```
This will cause the entire login session to crash. Even if it's unexpected, then you should probably use ```g_critical()``` ```return```

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+
+/* iterator function for update_machine() */
+static void
+update_machine_property(JsonObject *object,
```
Style: needs a space after function name.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+}
+
+static gboolean
+update_machine (const char *filename, const char* hostname, JsonNode *info, GError **error)
```
Style: arguments should be on different lines.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+      g_assert (JSON_NODE_HOLDS_OBJECT(cur_props));
+      json_object_foreach_member (json_node_get_object (info), update_machine_property, json_node_get_object (cur_props));
+    }
+  else
```
Style: Multiline block needs braces.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+
+      g_variant_get (parameters, "(&s&s@a{sv})", &filename, &hostname, &info_v);
+      info_json = json_gvariant_serialize (info_v);
+      /* g_debug ("Updating %s for machine %s: %s", filename, hostname, g_variant_print (info_v, TRUE)); */
```
g_debug() statements can remain uncommented. They're useful later with ```G_MESSAGES_DEBUG```

martinpitt replied:
I left this commented as it causes a memleak (`g_variant_print()` result needs to be freed and thus also needs a new variable) and it's relatively expensive. I'll uncomment it but leave out the value of the gvariant, it was mostly just necessary for developing the correct disassembling of the variant.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+
+  /* reset pending counter before we do any actual work, to avoid races */
+  pending_updates = 0;
+  machines = get_machines ();
```
I'm not against using a property for Machines ... but we should be using the ```invalidated``` parameter of PropertiesChanged instead of automatically doing this expensive operation each time ... even if nobody's listening.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+
+
+static void
+on_machines_changed (GFileMonitor     *monitor,
```
Style: Function arguments aren't lined up elsewhere in the C code.

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,451 @@
+  GVariant *machines, *signal_value;
+  GVariantBuilder builder;
+  GError *error = NULL;
+
```
We need to check if pending_updates is zero here. If it is then this should just return immediately without generating more PropertiesChanged events.

martinpitt replied:
`on_machines_changed()` already does that:
```
      if (pending_updates++ == 0)
        g_timeout_add (100, update_properties, user_data);
```
I can certainly add the check again here, but I'm afraid I'm missing what that would buy?

stefwalter replied:
Oh ok. As long as we're not scheduling multiple update_properties() timeouts.

martinpitt commented:
Pushed with addressing all of Stef's comments (only invalidating the property, stylistic fixes, g_error etc.)

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,468 @@
+                                 &error);
+  if (error != NULL)
+    {
+      g_critical ("failed to send PropertiesChanged signal: %s", error->message);
```
This can happen during shutdown when the DBus connection is closed. I don't think it's a g_critical() unless we account for this:

```
src/bridge/cockpitdbuscache.c:      g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CLOSED))
```

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,468 @@
+      g_error_free (error);
+      return;
+    }
+  g_signal_connect (machines_monitor, "changed", G_CALLBACK (on_machines_changed), connection);
```
Where is machines_monitor unreffed? Strange ... why doesn't it show up in memory leak info? I guess we're not covering cockpit-bridge startup/shutdown in our valgrind checks.

For completeness this could use a cockpit_dbus_machines_cleanup() function.

martinpitt replied:
FTR, as this isn't shown as "outdated" automatically: this is fixed now, I added a _cleanup().

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,468 @@
+
+          if (!JSON_NODE_HOLDS_VALUE (prop_node))
+            {
+              g_warning ("%s: host name definition %s: property %s does not contain a simple value, ignoring", path, hostname, propname);
```
This should be a g_message().

stefwalter commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,468 @@
+static guint pending_updates;
+
+static const char *
+machines_json_dir (void)
```
Isn't the expectation with XDG style conf dirs that we load all files from all the directories? This seems to choose the last directory and only operate on that. Am I wrong here (on either count)?

martinpitt replied:
Agreed, will rework it to do that (makes the globbing a bit more complicated, but *shrug*).

martinpitt replied:
Ah, it would not only affect the globbing, but we'd also need multiple file monitors and we would still need to pick one that we use for writing into. This is a lot of production code complication just to support tests, so as discussed on IRC, I'll move this back to the original proposal of a single dir env var, but call it `$COCKPIT_TEST_CONFIG_DIR` in line with the existing `$COCKPIT_TEST_SERVER_PORT`.

stefwalter commented on src/ws/test-server.c:
```diff
@@ -774,8 +776,19 @@ main (int argc,
   g_setenv ("XDG_DATA_HOME", SRCDIR "/src/bridge/mock-resource/home", TRUE);
   g_setenv ("XDG_DATA_DIRS", SRCDIR "/src/bridge/mock-resource/system", TRUE);
+  g_setenv ("XDG_CONFIG_DIRS", config_dir, TRUE);
+  g_setenv ("COCKPIT_DATA_DIR", data_dir, TRUE);
```
Is this variable added just for unit testing? If so I would rely on integration tests and not worry about yet more environment variables.

martinpitt replied:
Yes, just for verifying that `/var/lib/cockpit/machines.json` still gets read. But I can turn this into an integration test. I'm going to add one for adding the migration from /var to /etc anyway (as soon as this one lands), so after that this wouldn't add much extra overhead.

stefwalter commented on src/ws/test-server.c:
```diff
@@ -875,5 +888,13 @@ main (int argc,
 
+  /* clean up temporary config/data dirs */
+  rm_rf_argv[2] = config_dir;
+  g_spawn_sync(NULL, rm_rf_argv, NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL);
```
Space needed after function name.

stefwalter commented on src/ws/test-server.c:
```diff
@@ -875,5 +888,13 @@ main (int argc,
+  g_spawn_sync(NULL, rm_rf_argv, NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL);
+  g_free (config_dir);
+  rm_rf_argv[2] = data_dir;
+  g_spawn_sync(NULL, rm_rf_argv, NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL);
```
Space needed after function name.

martinpitt commented on src/bridge/cockpitdbusmachines.c:
```diff
@@ -0,0 +1,468 @@
+{
+  /* just log the error for debugging */
+  if (eerrno != ENOENT)
+    g_warning ("%s: cannot read: %s", epath, g_strerror (eerrno));
```
I think this actually should be a g_warning() as it's really unexpected that `glob()` itself fails -- it means that someone made /etc/cockpit/machines.d/ inaccessible, so this should really stand out IMHO.

martinpitt commented:
Re-pushed with addressing @stefwalter's points, thanks for the review! This now moves the handling of the old file in /var from a unit into an integration test which gets rid of `$COCKPIT_DATA_DIR`, does the warning → message  downgrade for most issues (except for `glob()` error, see above), unrefs the file monitor at bridge shutdown (busy work IMHO, but *shrug*), and goes back from multi-dir `$XDG_CONFIG_DIRS` to single-dir `$COCKPIT_TEST_CONFIG_DIR` (as above).

martinpitt commented:
The atomic failures seem real:
```
$ test/verify/check-docker-storage TestDockerStorage.testDevmapper
[...]
Error: failed to call cockpit.Machines.Update():  Failed to create file '/etc/cockpit/machines.d/99-webui.json.1T45VY': No such file or directory

$ test/verify/check-multi-machine-key
[...]
Error: failed to call cockpit.Machines.Update():  Failed to create file '/etc/cockpit/machines.d/99-webui.json.U2SNWY': No such file or directory
```
I'll fix those.

`test/verify/check-shutdown-restart` fails as well, I suppose for a similar reason.

stefwalter commented:
In addition to ```g_mkdir_with_parents``` this likely also needs ```/etc/cockpit/machines.d/``` to be owned by the cockpit-bridge subpackage, not cockpit-ws.

martinpitt commented:
Right. I need the directory pre-created in the package as otherwise we can't watch it for changes, and the (initially) unprivileged bridge can't create it either. I even mentioned that in the commit log, but forgot to put the new dir into the package, duh..

martinpitt commented:
Fixed with [this relative diff](https://paste.fedoraproject.org/paste/f1EoNPUSeQ-S8pnb-VIDOF5M1UNdIGYhyRLivL9gydE=), the fedora-atomic tests now pass for me locally.
